### ISSUE TITLE
Question, not a bug, but Github has no commenting or direct message feature :(

### ISSUE THREAD

**Issue by ArdRiCael** [JeffVi/Arduino-I2C-RW1063-0B-002#1]
Hi Jeff, 

I was trying to implement your code, but I wasn't receiving an output. I had a couple of questions:
How does the LCD know which character to print? I know it has a character input, how how does your code in the header convert this to the appropriate hex code? Furthermore, why is a command byte not used in the print functions? In the documentation, the DDRAM and CGRAM use address counters, which I cannot figure out. I'm very happy to move this dicussion elsewhere, I have no desire to clog up the issues page for this very-much appreciated repository. 

I'm using the MIDAS 2x8 LCD, which uses the RW1063. 

Best,
ArdRiCael

**Comment by JeffVi** [comment]
Hi @ArdRiCael,
Sorry for the wait, I was quite busy these days.
I'll try to answer your questions:
> How does the LCD know which character to print?
- There is no need to convert a character hex code, if you check the font table, you'll see that the CGRAM value for a character correspond to its ascii value (To write `a` you need to send `01100001`, or `0x61` which is the hex ascii code for `a`). If you plan on printing non-ascii characters, it will require updating the repo.
> Why is a command byte not used in the print functions?
- I assume you mean control byte(?). If so, [it is used](https://github.com/JeffVi/Arduino-I2C-RW1063-0B-002/blob/4a3dc9ff702c473bc3f9cacbee0dda18e22b3bbf/lib_I2CLCD.cpp#L152), but its value differs from other functions. It is set as `0x40` instead of `0x00` to indicate that the following bytes have to be stored in the display RAM.
> In the documentation, the DDRAM and CGRAM use address counters, which I cannot figure out.
- The `locateCursorLCD` function converts a `(x, y)` coordinate to the corresponding address counter. The display size is 4x20, so `x` varies from 0 to 19, and `y` varies from 0 to 3.

I hope this helps,
Jeff

**Comment by ArdRiCael** [comment]
Hi @JeffVi, 

No worries about the reply time, honestly this was a good turn around time, and very much appreciated!

Oh, the ascii thing is very handy! I understand the control byte now, thanks for that.  

> The locateCursorLCD function converts a (x, y) coordinate to the corresponding address counter. The display size is 4x20, so x varies from 0 to 19, and y varies from 0 to 3.

Does this mean that for the character print, the afformentioned CGRAM command should be OR'd with the cursor location address counter? 

I thought I'd found my issue (I'd forgotten the pull-up resistor for the I2C lines) but no dice. I'm going insane, I might just buy a new LCD, I cannot figure it out. It must be a wiring issue or a board issue, but I just can't find it. 

Best,
ArdRiCael

**Comment by JeffVi** [comment]
@ArdRiCael,
I think you are confusing CGRAM, DDRAM, and addresses.
I'll try to briefly explain those.
### CGRAM/DDRAM
Those are the RAM where is stored data. There is a specific command to edit said data. When using that command to edit RAM data, the data modified is located at the current CGRAM/DDRAM addresses. These addresses can be selected through a different command. To edit a value, you typically select the CGRAM/DDRAM address (with the relevant command), then edit the data (with the relevant command, which is not the same as the command to set the address). Note that after editing data, the address is automatically incremented (to select the next data).
#### CGRAM
Contains the patterns used to print characters on the display. The "shapes" of available characters. You typically don't need to edit it unless you want to create custom characters.
#### DDRAM
Contains the character codes that are printed on the display. When you call the `clearDisplayLCD` function, it sets all data to `0x20`, which is the ascii code for `space`.

### TL;DR
To print a character:
- You select a DDRAM address (if the current selected one is not correct)
- You send which character code should be printed
- The DDRAM is updated with that new data, and the DDRAM address is incremented
- The display prints the CGRAM pattern that corresponds to the character code of the DDRAM

I hope this will clarify stuff for you,
Jeff
